halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1357,6 +1573,38 @@ impl<F: Field> ConstraintSystem<F> {
         tmp
     }
 
+    /// Requests a challenge that is usable after the given phase.
+    pub fn challenge_usable_after(&mut self, phase: Phase) -> Challenge {
+        self.assert_no_phase_skipped(phase, "Challenge");
+        let tmp = Challenge {
+            index: self.num_challenges,
+            phase,
+        };
+        self.num_challenges += 1;
+        self.challenge_phase.push(phase);
+        tmp
+    }
+
+    /// Assert phase-aware resources are allocated in order, and avoids any
+    /// phase to be skipped accidentally to cause unexpected issue in the future.
+    fn assert_no_phase_skipped(&self, phase: Phase, resource: &'static str) {
+        if let Some(previous_phase) = match phase {
+            Phase::Second => Some(Phase::First),
+            Phase::Third => Some(Phase::Second),
+            _ => None,
```

Comment:
If a new phase is added later, this will silently cause it to not be checked. Use an explicit match here instead:
```suggestion
            Phase::First => None,
            Phase::Second => Some(Phase::First),
            Phase::Third => Some(Phase::Second),
```